lib/contracts.rb

```diff
@@ -242,108 +267,110 @@ def call(*args, &blk)
+      args.insert(-2, {})
+    elsif Hash === args_contracts[-1] && !args[-1].is_a?(Hash)
+      args << {}
     end
+  end
 
-    # Explicitly append options={} if Hash contract is present
-    if @has_options_contract
-      if @has_proc_contract && Hash === @args_contracts[-2] && !_args[-2].is_a?(Hash)
-        _args.insert(-2, {})
-        _args_size += 1
-      elsif Hash === @args_contracts[-1] && !_args[-1].is_a?(Hash)
-        _args << {}
-        _args_size += 1
-      end
+  def validate_arg(arg, contract, validator, arg_pos, args)
+    if validator && validator[arg]
+      true
+    else
+      Contract.failure_callback({
+        :arg => arg,
+        :contract => contract,
+        :class => klass,
+        :method => method,
+        :contracts => self,
+        :arg_pos => arg_pos,
+        :total_args => args.size
+      })
     end
+  end
 
-    # Loop forward validating the arguments up to the splat (if there is one)
-    (@splat_lower_index || _args_size).times do |i|
-      _arg = _args[i]
-      contract = @args_contracts[i]
-      validator = @args_validators[i]
-
-
-      unless validator && validator[_arg]
-        return unless Contract.failure_callback({
-          :arg => _arg,
-          :contract => contract,
-          :class => @klass,
-          :method => @method,
-          :contracts => self,
-          :arg_pos => i + 1,
-          :total_args => args_size
-        })
-      end
+  def call_with(this, *args, &blk)
+    args << blk if blk
+
+    # Explicitly append blk=nil if nil != Proc contract violation anticipated
+    maybe_append_block!(args, blk)
+
+    # Explicitly append options={} if Hash contract is present
+    maybe_append_options!(args, blk)
 
+    # Loop forward validating the arguments up to the splat (if there is one)
+    (args_contract_index || args.size).times do |i|
+      contract = args_contracts[i]
+      arg = args[i]
+      validator = args_validators[i]
+      return unless validate_arg(arg, contract, validator, i+1, args)
       if contract.is_a?(Contracts::Func)
-        _args[i] = Contract.new(@klass, _arg, *contract.contracts)
+        args[i] = Contract.new(klass, arg, *contract.contracts)
       end
     end
 
     # If there is a splat loop backwards to the lower index of the splat
     # Once we hit the splat in this direction set its upper index
     # Keep validating but use this upper index to get the splat validator.
-    if @splat_lower_index
-      splat_upper_index = @splat_lower_index
-      (_args_size - @splat_lower_index).times do |i|
-        _arg = _args[_args_size - 1 - i]
+    if args_contract_index
+      splat_upper_index = args_contract_index
+      (args.size - args_contract_index).times do |i|
+        arg = args[args.size - 1 - i]
 
-        if Contracts::Args === @args_contracts[contracts_size - 1 - i]
+        if Contracts::Args === args_contracts[args_contracts.size - 1 - i]
```

Comment:
It might be worth benchmarking a cached value of size vs calculating it each time as the value is used a few times. Ruby may internally optimize it or it may be negligible.